typegate/src/hooks.ts

```diff
@@ -0,0 +1,67 @@
+// Copyright Metatype OÜ under the Elastic License 2.0 (ELv2). See LICENSE.md for usage.
+
+import { parseGraphQLTypeGraph } from "./graphql/graphql.ts";
+import { TypeGraph, TypeGraphDS } from "./typegraph.ts";
+import { upgradeTypegraph } from "./typegraph/versions.ts";
+
+interface PushHandler {
+  (tg: TypeGraphDS, prev: TypeGraphDS | null): Promise<TypeGraphDS>;
+}
+
+interface InitHandler {
+  (tg: TypeGraph, prev: TypeGraph | null, sync: boolean): Promise<void>;
+}
+
+interface Hooks {
+  onPush: PushHandler[]; // before redis synchronization -- only once
+  onInit: InitHandler[]; // after redis sync -- run on each typegate instance
+}
+
+const hooks: Hooks = {
+  onPush: [
+    (tg) => Promise.resolve(upgradeTypegraph(tg)),
+    (tg) => Promise.resolve(parseGraphQLTypeGraph(tg)),
+  ],
+  onInit: [],
+};
+
+export function registerHook(when: "onPush", handler: PushHandler): void;
+export function registerHook(when: "onInit", handler: InitHandler): void;
+export function registerHook(
+  when: keyof Hooks,
+  handler: Hooks[keyof Hooks][number],
+): void {
+  (hooks[when] as unknown[]).push(handler);
+}
+
+const previousVersions: Map<string, TypeGraph> = new Map();
```

Comment:
Should not this come from the global register instead? This would avoid having typegraph stored at multiple places.

Reply:
Yes, that is right! This will not work at all since the next push might run on another instance.

The problem is that the hook might rewrite the typegraph:
- adding some computed data so it does get computed only once, before the redis synchronization;
- removing some extra data that was only needed for this hook;
- ...

Only the last version (throughout all the hooks) is stored in the global register and synchronized...

I was even thinking about storing multiple versions, one version per hook... but hard to synchronize...

Reply:
What if we do not compare to previous version? What workflow would not be possible using another strategy?

Reply:
It is just for optimization: for example skip the migrations if the migrations files did not change. We can drop it...